src/types.rs

```diff
@@ -8,33 +8,303 @@
+                                -> Option<String>
+    where I: Iterator<Item = &'a ast::PathSegment>
+{
+    let mut first = true;
+
+    for segment in iter {
+        let extra_offset = extra_offset(&buffer, offset);
+        let remaining_width = try_opt!(width.checked_sub(extra_offset));
+        let new_offset = offset + extra_offset;
+        let segment_string = try_opt!(rewrite_segment(segment,
+                                                      &mut span_lo,
+                                                      span_hi,
+                                                      context,
+                                                      remaining_width,
+                                                      new_offset));
+
+        if first {
+            first = false;
+        } else {
+            buffer.push_str("::");
+        }
+
+        buffer.push_str(&segment_string);
+    }
+
+    Some(buffer)
+}
+
+enum SegmentParam<'a> {
+    LifeTime(&'a ast::Lifetime),
+    Type(&'a ast::Ty),
+    Binding(&'a ast::TypeBinding),
+}
+
+impl<'a> SegmentParam<'a> {
+    fn get_span(&self) -> Span {
+        match *self {
+            SegmentParam::LifeTime(ref lt) => lt.span,
+            SegmentParam::Type(ref ty) => ty.span,
+            SegmentParam::Binding(ref binding) => binding.span,
+        }
+    }
+
+    fn get_string(&self) -> String {
+        match *self {
+            SegmentParam::LifeTime(ref lt) => pprust::lifetime_to_string(lt),
+            SegmentParam::Type(ref ty) => pprust::ty_to_string(ty),
+            SegmentParam::Binding(ref binding) => format!("{} = {}",
+                                                          binding.ident,
+                                                          pprust::ty_to_string(&*binding.ty)),
+        }
+    }
+}
+
+// This is a dirty hack to determine if we're in an expression or not. Generic
+// parameters are passed differently in expressions and items. We'd declare
+// a struct with Foo<A, B>, but call its functions with Foo::<A, B>::f().
+// We'd really rather not do this, but there doesn't seem to be an alternative
+// at this point.
+// FIXME: fails with spans containing comments with the characters < or :
+fn get_path_separator(codemap: &CodeMap,
+                      path_start: BytePos,
+                      segment_start: BytePos)
+                      -> &'static str {
+    let span = codemap::mk_sp(path_start, segment_start);
+    let snippet = codemap.span_to_snippet(span).unwrap();
+
+    for c in snippet.chars().rev() {
+        if c == ':' {
+            return "::"
+        } else if c.is_whitespace() || c == '<' {
+            continue;
+        } else {
+            return "";
+        }
+    }
+
+    unreachable!();
+}
+
+// Formats a path segment. There are some hacks involved to correctly determine
+// the segment's associated span since it's not part of the AST.
+//
+// The span_lo is guaranteed to be greater than the end of any previous
```

Comment:
Nit, do you mean that this function assumes that or guarantees that (I assume the former).